returnn/tf/util/data.py

```diff
@@ -998,12 +998,11 @@ def declare_same_as(self, other):
     if self.dyn_size is not None and other_same_base.dyn_size is not None:
```

Comment:
Can you move this whole block of code (`if self.dyn_size is not None and other_same_base.dyn_size is not None: ...`) a few lines up, just before the line `if self_same_as is not self: ...`?

This should make the error message and stacktrace more clear because then you still see the dim tags separated and not yet merged.

Reply:
The trace is pretty big but I think this is the most relevant part:
``` python
Finished training in epoch 2.
Run 3: Train results:
{1: EpochData(learningRate=0.01, error={
'dev_error': 0.6000000089406967,
'dev_score': 1.0964271708790534,
'train_error': 0.700000010430813,
'train_score': 1.0993032428518419,
}),
 2: EpochData(learningRate=0.01, error={
'dev_error': 0.6000000089406967,
'dev_score': 1.0963989421188671,
'train_error': 0.6600000098347665,
'train_score': 1.0985237666744554,
})}
Run 3: Forward cv seq 0:
[1 2 0 1 2]
Epoch 1, error key 'dev_error', current value 0.600000 vs prev value 0.600000, equal?
Epoch 1, error key 'dev_score', current value 1.096427 vs prev value 1.096427, equal?
Epoch 1, error key 'train_error', current value 0.700000 vs prev value 0.680000, equal?
EXCEPTION
Traceback (most recent call last):
  File "/Users/darkrock/returnn/tests/test_TFEngine.py", line 4406, in <module>
    line: v()
    locals:
      v = <local> <function test_rec_subnet_auto_optimize at 0x1658a9120>
  File "/Users/darkrock/returnn/tests/test_TFEngine.py", line 2603, in test_rec_subnet_auto_optimize
    line: run(run_idx=3, optimize_move_layers_out=True)
    locals:
      run = <local> <function test_rec_subnet_auto_optimize.<locals>.run at 0x168c6df30>
      run_idx = <not found>
      optimize_move_layers_out = <not found>
  File "/Users/darkrock/returnn/tests/test_TFEngine.py", line 2594, in test_rec_subnet_auto_optimize.<locals>.run
    line: numpy.testing.assert_almost_equal(error_value, prev_error_value, decimal=3)
    locals:
      numpy = <global> <module 'numpy' from '/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/numpy/__init__.py'>
      numpy.testing = <global> <module 'numpy.testing' from '/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/numpy/testing/__init__.py'>
      numpy.testing.assert_almost_equal = <global> <function assert_almost_equal at 0x165877ac0>
      error_value = <local> 0.700000010430813
      prev_error_value = <local> 0.6800000101327897
      decimal = <not found>
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/numpy/testing/_private/utils.py", line 599, in assert_almost_equal
    line: raise AssertionError(_build_err_msg())
    locals:
      AssertionError = <builtin> <class 'AssertionError'>
      _build_err_msg = <local> <function assert_almost_equal.<locals>._build_err_msg at 0x168c6de10>
AssertionError: 
Arrays are not almost equal to 3 decimals
 ACTUAL: 0.700000010430813
 DESIRED: 0.6800000101327897
```

Reply:
I don't exactly understand your comment. Did you do this change? I don't see it here?

Reply:
I did the change, didn't commit it yet. I'll push it so you can see the stacktrace in the pipeline.

Reply:
But I don't see this stacktrace/error you posted here in the pipeline?